synthma/condition_param.go

```diff
@@ -0,0 +1,65 @@
+package synthma
+
+import (
+	"errors"
+
+	"github.com/intervention-engine/fhir/models"
+	"github.com/intervention-engine/fhir/search"
+	"gopkg.in/mgo.v2/bson"
+)
+
+func init() {
+	// Register the condition-code parameter
+	search.GlobalRegistry().RegisterParameterInfo(ConditionCodeParamInfo)
+	search.GlobalRegistry().RegisterParameterParser(ConditionCodeParamInfo.Type, ConditionCodeParamParser)
+	search.GlobalMongoRegistry().RegisterBSONBuilder(ConditionCodeParamInfo.Type, ConditionCodeBSONBuilder)
+}
+
+// ConditionCodeParam represents the condition-code search parameter. Patient's may be searched
+// by conditions they have using a combination of system and code. This behaves exactly the same
+// as a standard TokenParam. See: http://hl7.org/fhir/2016Sep/search.html#token
+type ConditionCodeParam struct {
+	search.TokenParam
+}
+
+// ConditionCodeParamInfo represents the condition-code for Patients' conditions. This allows
+// patients to be searched by conditions they have.
+var ConditionCodeParamInfo = search.SearchParamInfo{
+	Resource: "Patient",
+	Name:     "condition-code",
+	Type:     "condition_code",
+}
+
+// ConditionCodeParamParser parses the parameter and returns a ConditionCodeParam.
+var ConditionCodeParamParser = func(info search.SearchParamInfo, data search.SearchParamData) (search.SearchParam, error) {
+	return &ConditionCodeParam{
+		TokenParam: *search.ParseTokenParam(data.Value, info),
+	}, nil
+}
+
+// ConditionCodeBSONBuilder builds the Mongo BSON object corresponding to the query by Patient condition.
+var ConditionCodeBSONBuilder = func(param search.SearchParam, searcher *search.MongoSearcher) (object bson.M, err error) {
+
+	cc, ok := param.(*ConditionCodeParam)
+	if !ok {
+		return nil, errors.New("Expected a ConditionCodeParam")
+	}
+
+	// First get a list of patient IDs for this condition
+	var conditions []models.Condition
+	if err := searcher.GetDB().C("conditions").Find(bson.M{"code.coding.code": cc.Code}).All(&conditions); err != nil {
+		return nil, err
+	}
```

Comment:
I think the above code could be made more efficient.  There are two things I'd suggest:

1) The search above is going to pull back all of the data for every matching condition (e.g., its id, code, effective dates, etc).  You only want the `subject.referenceid`, so all that other stuff is retrieved for no reason.  Use the [Query.Select](https://godoc.org/labix.org/v2/mgo#Query.Select) function to ask for just `subject.referenceid`.

2) Since you're only pulling back one piece of data, you don't even need to allocate full `Condition` structs.  You could create your own simple struct to hold the data, like this:
```go
var subjectWrappers []struct {
    Subject *models.Reference `bson:"subject,omitempty"`
}
if err := searcher.GetDB().C("conditions").Find(bson.M{"code.coding.code": cc.Code}).Select(bson.M{"subject.referenceid": 1}).All(&subjectWrappers); err != nil {
    return nil, err
}
```

The first suggestion is most important, but I'd suggest the second as well -- as it will save you memory.

One other thing to consider -- it's possible (although not likely in our case) that the subject references something other than `Patient`.  If you wanted to test for that, then you should select `subject` rather than `subject.referenceid`, and then check the type in the for loop below.